src/actions_ls.rs

```diff
@@ -212,7 +228,7 @@ impl ActionHandler {
         let vfs: &Vfs = &self.vfs;
         let result: Vec<CompletionItem> = panic::catch_unwind(move || {
             let pos = adjust_vscode_pos_for_racer(params.position);
-            let file_path = &params.textDocument.file_name();
+            let file_path = &uri_string_to_file_name(&params.text_document.uri);
```

Comment:
Since the spec says that the `uri` field should be an actual URI (that is just represented as a string in the JSON) it might be simpler to just make the field hold an `URI` (or URL I guess) directly. It would cause serde de-serialization to fail when interacting with a non-conforming implementation which may be annoying (but it would be correct).

In my opinion, since there will always be differing opinions on which methods to add to the types I'd prefer to have no methods or just the bare minimum number of methods lest the API is likely going to end up bloated by "wouldn't it be nice if..." methods. (I am a bit skeptic towards adding the constructor methods actually as they will basically be redundant once https://github.com/rust-lang/rust/issues/37340 is implemented and gets to stable).